### PR TITLE
Do not rely on the X-<protocol>-User HTTP header to detect a protocol

### PR DESCRIPTION
**Pull request by lpoulain**
## Description

Change the logic to detect a non-Trino protocol to stop relying on the X-<protocol>-User HTTP header.

## Additional context and related issues

Trino detects alternative protocols (e.g. Presto) by looking whether the X-<protocol>-User HTTP header exists. In some cases however no such header is provided. So change the logic to avoid relying on that particular HTTP header and instead look for any X-<protocol>-* HTTP header.

## Release notes

(X) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by findepi — APPROVED** [review]

**Comment by lpoulain** [comment]
In Galaxy's case for example

File: client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java [hunk 1]
```diff
@@ -190,8 +191,11 @@ public static ProtocolHeaders detectProtocol(Optional<String> alternateHeaderNam
         requireNonNull(headerNames, "headerNames is null");
 
         if (alternateHeaderName.isPresent() && !alternateHeaderName.get().equalsIgnoreCase("Trino")) {
-            if (headerNames.contains("X-" + alternateHeaderName.get() + "-User")) {
-                if (headerNames.contains("X-Trino-User")) {
+            String headerPrefix = "x-" + alternateHeaderName.get().toLowerCase(ENGLISH);
+            if (headerNames.stream()
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java:195]: Nit: these would look better without wrapping

**Comment by electrum** [comment]
Now that we do user mapping for authentication, it's legitimate for a client not to send the user header, and apparently some clients do not always send it.

File: client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java (hunk 1 above)
Comment (lpoulain) [reply client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java:195]: You mean something like:

```
            if (headerNames.stream().anyMatch(header -> header.toLowerCase(ENGLISH).startsWith(headerPrefix))) {
                if (headerNames.stream().anyMatch(header -> header.toLowerCase(ENGLISH).startsWith("x-trino-"))) {
```

**Comment by mosabua** [comment]
Can you clarify what the next here should be @electrum ?

File: client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java (hunk 1 above)
Comment (electrum) [reply client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java:195]: Yep, exactly

File: client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java [hunk 2]
```diff
@@ -190,8 +191,11 @@ public static ProtocolHeaders detectProtocol(Optional<String> alternateHeaderNam
...
+                    .anyMatch(header -> header.toLowerCase(ENGLISH).startsWith(headerPrefix))) {
+                if (headerNames.stream()
+                        .anyMatch(header -> header.toLowerCase(ENGLISH).startsWith("x-trino-"))) {
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java:198]:
```suggestion
            if (headerNames.stream().anyMatch(header -> header.toLowerCase(ENGLISH).startsWith(headerPrefix))) {
                if (headerNames.stream().anyMatch(header -> header.toLowerCase(ENGLISH).startsWith("x-trino-"))) {
```

**Pull request merged**